test/stdlib/OptionalBridge.swift

```diff
@@ -0,0 +1,135 @@
+    expectEqual(xx, yy)
+  case (.none, .none):
+    return
+  default:
+    expectUnreachable("\(T.self)??? values don't match: \(x) vs. \(y)")
+  }
+}
+
+tests.test("wrapped value") {
+  let unwrapped = "foo"
+  let wrapped = Optional(unwrapped)
+  let doubleWrapped = Optional(wrapped)
+
+  let unwrappedBridged = unwrapped as AnyObject
+  let wrappedBridged = wrapped as AnyObject
+  let doubleWrappedBridged = doubleWrapped as AnyObject
+  expectTrue(unwrappedBridged.isEqual(wrappedBridged)
+             && wrappedBridged.isEqual(doubleWrappedBridged))
+
+  let unwrappedCastBack = cast(unwrappedBridged, to: String.self)
+  let wrappedCastBack = cast(wrappedBridged, to: Optional<String>.self)
+  let doubleWrappedCastBack = cast(doubleWrappedBridged, to: Optional<String?>.self)
+
+  expectEqual(unwrapped, unwrappedCastBack)
+  expectEqual(wrapped, wrappedCastBack)
+  expectEqual(doubleWrapped, doubleWrappedCastBack)
+}
+
+tests.test("wrapped class instance") {
+  let unwrapped = LifetimeTracked(0)
+  let wrapped = Optional(unwrapped)
+
+  expectTrue(wrapped as AnyObject === unwrapped as AnyObject)
+}
+
+tests.test("nil") {
+  let null: String? = nil
+  let wrappedNull = Optional(null)
+  let doubleWrappedNull = Optional(wrappedNull)
+
+  let nullBridged = null as AnyObject
+  let wrappedNullBridged = wrappedNull as AnyObject
+  let doubleWrappedNullBridged = doubleWrappedNull as AnyObject
+
+  expectTrue(nullBridged === NSNull())
+  expectTrue(wrappedNullBridged === NSNull())
+  expectTrue(doubleWrappedNullBridged === NSNull())
+
+  let nullCastBack = cast(nullBridged, to: Optional<String>.self)
+  let wrappedNullCastBack = cast(nullBridged, to: Optional<String?>.self)
+  let doubleWrappedNullCastBack = cast(nullBridged, to: Optional<String??>.self)
+
+  expectEqual(nullCastBack, null)
+  expectEqual(wrappedNullCastBack, wrappedNull)
+  expectEqual(doubleWrappedNullCastBack, doubleWrappedNull)
+}
+
+tests.test("nil in nested optional") {
+  let doubleNull: String?? = nil
+  let wrappedDoubleNull = Optional(doubleNull)
+
+  let doubleNullBridged = doubleNull as AnyObject
+  let wrappedDoubleNullBridged = wrappedDoubleNull as AnyObject
+
+  expectTrue(doubleNullBridged === wrappedDoubleNullBridged)
+  expectTrue(doubleNullBridged !== NSNull())
+
+  let doubleNullCastBack = cast(doubleNullBridged, to: Optional<String?>.self)
+  let wrappedDoubleNullCastBack = cast(doubleNullBridged, to: Optional<String??>.self)
+
+  expectEqual(doubleNullCastBack, doubleNull)
+  expectEqual(wrappedDoubleNullCastBack, wrappedDoubleNull)
+
+  let tripleNull: String??? = nil
+  let tripleNullBridged = tripleNull as AnyObject
+
+  expectTrue(doubleNullBridged !== tripleNullBridged)
+
+  let tripleNullCastBack = cast(tripleNullBridged, to: Optional<String??>.self)
+  expectEqual(tripleNullCastBack, tripleNull)
+}
+
+tests.test("collection of Optional") {
+  let holeyArray: [LifetimeTracked?] = [LifetimeTracked(0), nil, LifetimeTracked(1)]
```

Comment:
Does the reverse work? `someNSArray as [LifetimeTracked?]`

This pattern if ever used will never really be portable to swift-corelibs-foundation which is a bit unfortunate.